File: app/src/prod/java/ch/zuehlke/sbb/reddit/Injection.java
package ch.zuehlke.sbb.reddit;


import android.content.Context;
import android.support.annotation.NonNull;

import com.google.gson.Gson;
import com.google.gson.GsonBuilder;
import com.google.gson.reflect.TypeToken;

import java.lang.reflect.Modifier;
import java.lang.reflect.Type;
import java.util.List;

import ch.zuehlke.sbb.reddit.data.source.RedditRepository;
import ch.zuehlke.sbb.reddit.data.source.local.RedditNewsLocalDataSource;
import ch.zuehlke.sbb.reddit.data.source.remote.RedditAPI;
import ch.zuehlke.sbb.reddit.data.source.remote.RedditNewsDataRemoteDataSource;
import ch.zuehlke.sbb.reddit.data.source.remote.model.posts.RedditPostElement;
import retrofit2.Retrofit;
import retrofit2.converter.gson.GsonConverterFactory;

import static ch.zuehlke.sbb.reddit.data.source.remote.RedditElementTypeAdapterFactory.getElementTypeAdapterFactory;
import static com.google.common.base.Preconditions.checkNotNull;

/**
 * Enables injection of production implementations for
 * {@link ch.zuehlke.sbb.reddit.data.source.RedditDataSource} at compile time.
 */
public class Injection {

    private static final String REDDIT_END_POINT = "https://www.reddit.com/r/dota2/";

    private static RedditAPI redditAPI;
    private static Retrofit retrofit;

    public static final Type type = new TypeToken<List<RedditPostElement>>() {
    }.getType();

    public static final Gson gson = new GsonBuilder()
            .registerTypeAdapterFactory(getElementTypeAdapterFactory())
            .create();

    public static RedditRepository provideRedditNewsRepository(@NonNull Context context) {
         checkNotNull(context);
        return RedditRepository.getInstance(RedditNewsDataRemoteDataSource.getInstance(context,getRedditAPI(getRetroFit())),
                RedditNewsLocalDataSource.getInstance(context));
    }

    public static RedditAPI getRedditAPI(Retrofit retrofit){
        if(redditAPI ==null){
            redditAPI = retrofit.create(RedditAPI.class);
        }
        return redditAPI;
    }

    public static Retrofit getRetroFit(){
        if(retrofit == null){
            Gson gson = new GsonBuilder()
                    .excludeFieldsWithModifiers(Modifier.FINAL, Modifier.TRANSIENT, Modifier.STATIC)
                    .excludeFieldsWithoutExposeAnnotation()
                    .registerTypeAdapterFactory(getElementTypeAdapterFactory())
                    .create();

            retrofit = new Retrofit.Builder()
                    .baseUrl(REDDIT_END_POINT)
                    .addConverterFactory(GsonConverterFactory.create(gson))
                    .build();
        }
        return retrofit;
    }
}

File: app/src/prod/java/ch/zuehlke/sbb/reddit/Injection.kt
package ch.zuehlke.sbb.reddit


import android.content.Context
import ch.zuehlke.sbb.reddit.data.source.RedditRepository
import ch.zuehlke.sbb.reddit.data.source.RemoteDataMapper
import ch.zuehlke.sbb.reddit.data.source.local.RedditNewsLocalDataSource
import ch.zuehlke.sbb.reddit.data.source.remote.RedditAPI
import ch.zuehlke.sbb.reddit.data.source.remote.RedditElementTypeAdapterFactory.Companion.elementTypeAdapterFactory
import ch.zuehlke.sbb.reddit.data.source.remote.RedditNewsDataRemoteDataSource
import ch.zuehlke.sbb.reddit.data.source.remote.model.posts.RedditPostElement
import com.google.common.base.Preconditions.checkNotNull
import com.google.gson.GsonBuilder
import com.google.gson.reflect.TypeToken
import retrofit2.Retrofit
import retrofit2.converter.gson.GsonConverterFactory
import java.lang.reflect.Modifier

/**
 * Enables injection of production implementations for
 * [ch.zuehlke.sbb.reddit.data.source.RedditDataSource] at compile time.
 */
object Injection {

    private val REDDIT_END_POINT = "https://www.reddit.com/r/kotlin/"

    private var redditAPI: RedditAPI? = null
    private var retrofit: Retrofit? = null

    val type = object : TypeToken<List<RedditPostElement>>() {

    }.type


    val gson = GsonBuilder()
            .registerTypeAdapterFactory(elementTypeAdapterFactory)
            .create()

    val remoteDataMapper = RemoteDataMapper.getInstance(gson, type)

    fun provideRedditNewsRepository(context: Context): RedditRepository {
        checkNotNull(context)
        return RedditRepository.getInstance(RedditNewsDataRemoteDataSource.getInstance(context, getRedditAPI(retroFit), remoteDataMapper),
                RedditNewsLocalDataSource.getInstance(context), context)
    }

    fun getRedditAPI(retrofit: Retrofit): RedditAPI {
        if (redditAPI == null) {
            redditAPI = retrofit.create<RedditAPI>(RedditAPI::class.java!!)
        }
        return redditAPI!!
    }

    val retroFit: Retrofit
        get() {
            if (retrofit == null) {
                val gson = GsonBuilder()
                        .excludeFieldsWithModifiers(Modifier.FINAL, Modifier.TRANSIENT, Modifier.STATIC)
                        .excludeFieldsWithoutExposeAnnotation()
                        .registerTypeAdapterFactory(elementTypeAdapterFactory)
                        .create()

                retrofit = Retrofit.Builder()
                        .baseUrl(REDDIT_END_POINT)
                        .addConverterFactory(GsonConverterFactory.create(gson))
                        .build()
            }
            return retrofit!!
        }
}
